Numerical instability for expm1 with complex arguments

The output of `np.expm1` becomes unstable for small values (around 10^{-15} to 10^{-16}) when the argument is complex. This does not seem to be an issue with the numerical precision associated with the real and imaginary components: using a `float32` data type works as expected. 

The problem arises because of the way `expm1` is [currently implemented](https://github.com/numpy/numpy/blob/deea4983aedfa96905bbaee64e3d1de84144303f/numpy/core/src/umath/funcs.inc.src#L360-L367) for complex arguments. In particular:

```c
static void
nc_expm1@c@(@ctype@ *x, @ctype@ *r)
{
    @ftype@ a = npy_exp@c@(x->real);
    // If x->imag == 0, then we evaluate exp(x->real) - 1 naively 
    // rather than using a more stable implementation
    r->real = a*npy_cos@c@(x->imag) - 1.0@c@;
    r->imag = a*npy_sin@c@(x->imag);
    return;
}
```

### Reproducing code example:

```python
import numpy as np
from matplotlib import pyplot as plt

fig, ax = plt.subplots()

x = np.logspace(-16.2, -14)
xs = [
    x,
    x.astype(np.complex),
    x.astype(np.float32),
]

ys = [np.expm1(x_) for x_ in xs]

for x, y in zip(xs, ys):
    ax.plot(x, y, marker='.', label=f'{x.dtype}: np.expm(x)')

ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlabel('x')
ax.legend(loc='upper left')

fig.tight_layout()
```

Virtually identical behaviour can be observed by replacing `ys = [np.expm1(x_) for x_ in xs]` with `ys = [-np.expm1(-x_) for x_ in xs]`, i.e. the problem is not unique to positive values of the argument.

![image](https://user-images.githubusercontent.com/966348/61252620-025a2900-a756-11e9-910d-8b06c7d9221d.png)


### Error message:

No error message to report. However, the expected behaviour would be for the complex datatype to behave in the same fashion as the float32 and float64 data types.

### Numpy/Python version information:

```
1.16.4 3.7.2 (default, Jul 11 2019, 09:18:04) 
[Clang 10.0.1 (clang-1001.0.46.4)]
```

Yes, I believe the issue is related to the [specific implementation](https://github.com/numpy/numpy/blob/deea4983aedfa96905bbaee64e3d1de84144303f/numpy/core/src/umath/funcs.inc.src#L360-L367) of `expm1` for complex arguments within numpy:

```c
static void
nc_expm1@c@(@ctype@ *x, @ctype@ *r)
{
    @ftype@ a = npy_exp@c@(x->real);
    // If x->imag == 0, then we evaluate exp(x->real) - 1 naively 
    // rather than using a more stable implementation
    r->real = a*npy_cos@c@(x->imag) - 1.0@c@;
    r->imag = a*npy_sin@c@(x->imag);
    return;
}
```

Note that `scipy.special.expm1` has the expected behaviour. Maybe we can reuse said implementation if the license allows it.

Example to reproduce the same issue on Linux:

```python
# test.py
import sys, numpy as np

print(np.__version__, sys.version)

assert np.expm1(1e-17) > 0, "failed for real number"
assert np.expm1(1e-17 + 0j) > 0, "failed for complex number"
```

```Dockerfile
# Dockerfile
FROM python:3.6.8
RUN pip install numpy
COPY test.py .
```

```bash
# script to run example
$ docker build -t numpy-test .
[...]
Successfully tagged numpy-test:latest
$ docker run --rm numpy-test python test.py
1.16.4 3.6.8 (default, Mar 27 2019, 08:49:59) 
[GCC 6.3.0 20170516]
Traceback (most recent call last):
  File "test.py", line 6, in <module>
    assert np.expm1(1e-17 + 0j) > 0, "failed for complex number"
AssertionError: failed for complex number
```

--- a/numpy/core/src/umath/funcs.inc.src
+++ b/numpy/core/src/umath/funcs.inc.src
@@ -360,9 +360,9 @@ nc_exp2@c@(@ctype@ *x, @ctype@ *r)
 static void
 nc_expm1@c@(@ctype@ *x, @ctype@ *r)
 {
-    @ftype@ a = npy_exp@c@(x->real);
-    r->real = a*npy_cos@c@(x->imag) - 1.0@c@;
-    r->imag = a*npy_sin@c@(x->imag);
+    @ftype@ a = npy_cos@c@(x->imag);
+    r->real = npy_expm1@c@(x->real) * a + (a - 1);
+    r->imag = npy_exp@c@(x->real) * npy_sin@c@(x->imag);
     return;
 }
 

--- a/numpy/core/src/umath/funcs.inc.src
+++ b/numpy/core/src/umath/funcs.inc.src
@@ -360,8 +360,8 @@ nc_exp2@c@(@ctype@ *x, @ctype@ *r)
 static void
 nc_expm1@c@(@ctype@ *x, @ctype@ *r)
 {
-    @ftype@ a = npy_cos@c@(x->imag);
-    r->real = npy_expm1@c@(x->real) * a + (a - 1);
+    @ftype@ a = npy_sin@c@(x->imag / 2);
+    r->real = npy_expm1@c@(x->real) * npy_cos@c@(x->imag) - 2 * a * a;
     r->imag = npy_exp@c@(x->real) * npy_sin@c@(x->imag);
     return;
 }

--- a/numpy/core/tests/test_umath.py
+++ b/numpy/core/tests/test_umath.py
@@ -888,6 +888,12 @@ class TestExpm1(object):
         assert_equal(ncu.expm1(np.inf), np.inf)
         assert_equal(ncu.expm1(-np.inf), -1.)
 
+    def test_complex(self):
+        x = np.asarray(1e-12)
+        assert_allclose(x, ncu.expm1(x))
+        x = x.astype(np.complex128)
+        assert_allclose(x, ncu.expm1(x))
+
 
 class TestHypot(object):
     def test_simple(self):

